### PR TITLE
Fix keepalive for large jumps on tabs and lists.

### LINKED ISSUE #11895: _debugUltimatePreviousSiblingOf(after, equals: _firstChild) is not true.
## Steps to Reproduce

I have a PageView that contains the top level widgets for my app, with switching between them controlled by a BottomNavBar. As per @Hixie's advice in the chat, I am adding ```AutomaticKeepAliveClientMixin``` to these top level widgets so the PageView won't recycle them and will preserve state. After I added this to the main widget that contains a network loaded ListView, on occasion when switching back to this ListView from another tab I get this failed assertion in the framework.

## Logs

```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following assertion was thrown building NotificationListener<KeepAliveNotification>:
'package:flutter/src/rendering/object.dart': Failed assertion: line 3031 pos 14:
'_debugUltimatePreviousSiblingOf(after, equals: _firstChild)': is not true.
Either the assertion indicates an error in the framework itself, or we should provide substantially
more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new
When the exception was thrown, this was the stack:
#2      RenderSliver&ContainerRenderObjectMixin._insertIntoChildList (package:flutter/src/rendering/object.dart:3031:14)
#3      RenderSliver&ContainerRenderObjectMixin.insert (package:flutter/src/rendering/object.dart:3065:5)
#4      RenderSliverMultiBoxAdaptor.insert (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:193:17)
#5      SliverMultiBoxAdaptorElement.insertChildRenderObject (package:flutter/src/widgets/sliver.dart:840:18)
#6      RenderObjectElement.attachRenderObject (package:flutter/src/widgets/framework.dart:4380:35)
#7      RenderObjectElement.mount (package:flutter/src/widgets/framework.dart:4151:5)
#8      SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:4515:16)
#9      Element.inflateWidget (package:flutter/src/widgets/framework.dart:2882:14)
#10     Element.updateChild (package:flutter/src/widgets/framework.dart:2685:12)
#11     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#12     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#13     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3514:5)
#14     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3509:5)
#15     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2882:14)
#16     Element.updateChild (package:flutter/src/widgets/framework.dart:2685:12)
#17     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#18     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#19     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3514:5)
#20     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3509:5)
#21     ParentDataElement.mount (package:flutter/src/widgets/framework.dart:3827:16)
#22     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2882:14)
#23     Element.updateChild (package:flutter/src/widgets/framework.dart:2685:12)
#24     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#25     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#26     ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:3514:5)
#27     StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:3652:22)
#28     ComponentElement.mount (package:flutter/src/widgets/framework.dart:3509:5)
#29     Element.inflateWidget (package:flutter/src/widgets/framework.dart:2882:14)
#30     Element.updateChild (package:flutter/src/widgets/framework.dart:2685:12)
#31     SliverMultiBoxAdaptorElement.performRebuild (package:flutter/src/widgets/sliver.dart:696:34)
#32     SliverMultiBoxAdaptorElement.update (package:flutter/src/widgets/sliver.dart:665:7)
#33     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#34     RenderObjectElement.updateChildren (package:flutter/src/widgets/framework.dart:4250:32)
#35     MultiChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:4631:17)
#36     _ViewportElement.update (package:flutter/src/widgets/viewport.dart:150:17)
#37     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#38     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#39     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#40     ProxyElement.update (package:flutter/src/widgets/framework.dart:3781:5)
#41     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#42     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:4523:14)
#43     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#44     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:4523:14)
#45     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#46     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:4523:14)
#47     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#48     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#49     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#50     StatefulElement.update (package:flutter/src/widgets/framework.dart:3671:5)
#51     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#52     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#53     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#54     StatefulElement.update (package:flutter/src/widgets/framework.dart:3671:5)
#55     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#56     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#57     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#58     StatelessElement.update (package:flutter/src/widgets/framework.dart:3596:5)
#59     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#60     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#61     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#62     StatefulElement.update (package:flutter/src/widgets/framework.dart:3671:5)
#63     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#64     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#65     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#66     ProxyElement.update (package:flutter/src/widgets/framework.dart:3781:5)
#67     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#68     RenderObjectElement.updateChildren (package:flutter/src/widgets/framework.dart:4250:32)
#69     MultiChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:4631:17)
#70     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#71     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#72     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#73     ProxyElement.update (package:flutter/src/widgets/framework.dart:3781:5)
#74     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#75     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#76     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#77     StatefulElement.update (package:flutter/src/widgets/framework.dart:3671:5)
#78     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#79     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:4523:14)
#80     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#81     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#82     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#83     StatelessElement.update (package:flutter/src/widgets/framework.dart:3596:5)
#84     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#85     SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:4523:14)
#86     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#87     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#88     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#89     StatefulElement.update (package:flutter/src/widgets/framework.dart:3671:5)
#90     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#91     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#92     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#93     StatefulElement.update (package:flutter/src/widgets/framework.dart:3671:5)
#94     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#95     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#96     Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#97     ProxyElement.update (package:flutter/src/widgets/framework.dart:3781:5)
#98     Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#99     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#100    Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#101    ProxyElement.update (package:flutter/src/widgets/framework.dart:3781:5)
#102    Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#103    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#104    Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#105    StatefulElement.update (package:flutter/src/widgets/framework.dart:3671:5)
#106    Element.updateChild (package:flutter/src/widgets/framework.dart:2674:15)
#107    ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:3546:16)
#108    Element.rebuild (package:flutter/src/widgets/framework.dart:3435:5)
#109    BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2260:33)
#110    BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:503:20)
#111    BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:189:5)
#112    BindingBase&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:695:15)
#113    BindingBase&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:643:9)
#114    _invoke (file:///b/build/slave/Mac_Engine/build/src/flutter/lib/ui/hooks.dart:86)
#115    _drawFrame (file:///b/build/slave/Mac_Engine/build/src/flutter/lib/ui/hooks.dart:75)
(elided 2 frames from class _AssertionError)
════════════════════════════════════════════════════════════════════════════════════════════════════
Another exception was thrown: 'package:flutter/src/rendering/object.dart': Failed assertion: line 3031 pos 14: '_debugUltimatePreviousSiblingOf(after, equals: _firstChild)': is not true.
Another exception was thrown: 'package:flutter/src/rendering/object.dart': Failed assertion: line 3031 pos 14: '_debugUltimatePreviousSiblingOf(after, equals: _firstChild)': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
Another exception was thrown: 'package:flutter/src/widgets/framework.dart': Failed assertion: line 3437 pos 14: 'owner._debugCurrentBuildTarget == this': is not true.
```

```
Analyzing /Users/matt/code/flutter/kudos...
  lint • Don't import implementation files from another package at lib/feed.dart:5:8 • implementation_imports
  lint • Close instances of `dart.core.Sink` at lib/feed.dart:26:26 • close_sinks
  lint • Don't explicitly initialize variables to null at lib/loading_list_view.dart:23:5 • avoid_init_to_null
  lint • Don't import implementation files from another package at lib/post.dart:3:8 • implementation_imports
  hint • Unused import at lib/application.dart:4:8 • unused_import
  hint • Unused import at lib/feed.dart:2:8 • unused_import
  hint • Unused import at lib/feed.dart:5:8 • unused_import
  hint • Unused import at lib/feed.dart:7:8 • unused_import
  hint • Unused import at lib/feed.dart:12:8 • unused_import
  hint • Unused import at lib/feed.dart:13:8 • unused_import
  hint • This method overrides a method annotated as @mustCallSuper in 'State', but does not invoke the overridden method at lib/feed.dart:29:8 • must_call_super
  hint • This method overrides a method annotated as @mustCallSuper in 'State', but does not invoke the overridden method at lib/feed.dart:40:8 • must_call_super
  hint • The type parameter 'widgetAdapter' is annotated with @required but only named parameters without default value can be annotated with it at lib/loading_list_view.dart:23:5 • invalid_required_param
  hint • Unused import at lib/main.dart:6:8 • unused_import
  hint • The value of the field '_counter' isn't used at lib/main.dart:56:7 • unused_field
  hint • This method overrides a method annotated as @mustCallSuper in 'State', but does not invoke the overridden method at lib/main.dart:63:8 • must_call_super
  hint • This method overrides a method annotated as @mustCallSuper in 'State', but does not invoke the overridden method at lib/main.dart:71:8 • must_call_super
  hint • The method '_incrementCounter' isn't used at lib/main.dart:76:8 • unused_element
  hint • Unused import at lib/post.dart:7:8 • unused_import
  hint • This class inherits from a class marked as @immutable, and therefore should be immutable (all instance fields must be final) at lib/post.dart:10:7 • must_be_immutable
  hint • This class inherits from a class marked as @immutable, and therefore should be immutable (all instance fields must be final) at lib/post.dart:34:7 • must_be_immutable
  hint • The value of the local variable 'plainText' isn't used at lib/post.dart:71:10 • unused_local_variable
  hint • This class inherits from a class marked as @immutable, and therefore should be immutable (all instance fields must be final) at lib/post.dart:112:7 • must_be_immutable
  hint • This class inherits from a class marked as @immutable, and therefore should be immutable (all instance fields must be final) at lib/post.dart:160:7 • must_be_immutable
  hint • This class inherits from a class marked as @immutable, and therefore should be immutable (all instance fields must be final) at lib/profile.dart:10:7 • must_be_immutable
  hint • This method overrides a method annotated as @mustCallSuper in 'State', but does not invoke the overridden method at lib/profile.dart:27:10 • must_call_super
  hint • This class inherits from a class marked as @immutable, and therefore should be immutable (all instance fields must be final) at lib/profile.dart:145:7 • must_be_immutable
  hint • The class '_KudosEarned' isn't used at lib/profile.dart:218:7 • unused_element
  hint • Unused import at lib/send_photo.dart:5:8 • unused_import
  hint • Unused import at lib/usersearch.dart:6:8 • unused_import
  hint • Unused import at lib/usersearch.dart:7:8 • unused_import
  hint • This method overrides a method annotated as @mustCallSuper in 'State', but does not invoke the overridden method at lib/usersearch.dart:25:10 • must_call_super
32 issues found.
```

## Flutter Doctor

```
[✓] Flutter (on Mac OS X 10.12.6 16G29, locale en-US, channel alpha)
    • Flutter at /Users/matt/code/flutter/flutter
    • Framework revision b156a0f054 (3 weeks ago), 2017-08-11 22:01:40 -0700
    • Engine revision fef7d827d6
    • Tools Dart version 1.25.0-dev.9.0

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at /Users/matt/code/android/android
    • Platform android-25, build-tools 25.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 8.3.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.3.3, Build version 8E3004b
    • ios-deploy 1.9.1
    • CocoaPods version 1.3.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Ultimate Edition (version 2017.2.2)
    • Flutter plugin version 16.0
    • Dart plugin version 172.3757.55

[✓] Connected devices
    • iPhone 6s • 3DF999CB-1CD6-4649-8187-DDEFBDB14DFA • ios • iOS 10.3 (simulator)

```

> For more information about diagnosing and reporting Flutter bugs, please see [https://flutter.io/bug-reports/](https://flutter.io/bug-reports/).

### PR DESCRIPTION
**Pull request by dnfield**
Fixes #11895.

Successor to #21207 

The SliverMultiBoxAdaptorElement tries to keep track of the child it should be inserting children after when it needs to do an insertion. This breaks if the child it thinks it wants doesn't know what siblings it should have because it got thrown into the keepAliveBucket at some point - the insertion logic expects children to be connected and in the proper indexing order.

This pull makes sure that if the child you want to insert after is in the keep alive bucket, it's removed and reinserted to properly get connected to it's siblings.  This case happens in, for example, a multi tab view where the first tab is kept alive and you immediately go to tab 4 (but not if you go from 1 to 2).  It also manifests in a sliver list where you fling and then rebuild the list.

### CODE REVIEW

**Comment by marekchen** [comment]
@dnfield Does your pr solve this issue  #18832

**Comment by dnfield** [comment]
@marekchen - no.  The bug you're reporting there is not related to this issue.

**Comment by dnfield** [comment]
/cc @goderbauer - I'm seeing behavior occasionally where code added in https://github.com/flutter/flutter/pull/16445 is getting called after `context` is null (presumably because by the time the frame has been updated the widget is no longer in the tree).  Added some code here to check for nulls in these cases.

This should still be ok because of the code I introduced in sliver.dart, which should correct the bad parent data a bit later in the whole process if necessary.

File: packages/flutter/lib/src/widgets/automatic_keep_alive.dart [hunk 1]
```diff
@@ -89,6 +89,10 @@ class _AutomaticKeepAliveState extends State<AutomaticKeepAlive> {
         // build of this subtree. Wait until the end of the frame to update
         // the child when the child is guaranteed to be present.
         SchedulerBinding.instance.addPostFrameCallback((Duration timeStamp) {
+          // The element is no longer alive
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/automatic_keep_alive.dart]: This seems reasonable. When writing the code we were actually wondering if that case could ever happen and thought it wouldn't....

Can you add a test for this case?

File: packages/flutter/test/material/tabs_test.dart [hunk 2]
```diff
@@ -1754,4 +1771,54 @@ void main() {
...
+          ),
+        ),
+      ),
+    );
+    expect(find.text(AlwaysKeepAliveWidget.text), findsOneWidget);
+    await tester.tap(find.text('Tab4'));
+    await tester.pumpAndSettle();
+    await tester.pump();
+    expect(controller.index, 3);
+    expect(find.text('4'), findsOneWidget);
```
Comment (goderbauer) [inline packages/flutter/test/material/tabs_test.dart]: maybe assert at the beginning that text('4') didn't find anything to ensure that the setup is and stays correct? And assert at the end that AlwaysKeepAlive still finds something?

File: packages/flutter/test/widgets/list_view_test.dart [hunk 3]
```diff
@@ -86,12 +140,12 @@ void main() {
...
-                return new Container(
-                  child: new Text('$i'),
-                );
-              }
+                builder: (BuildContext context) {
+                  log.add(i);
+                  return new Container(
+                    child: new Text('$i'),
+                  );
+                }
```
Comment (goderbauer) [inline packages/flutter/test/widgets/list_view_test.dart]: Revert the whitespace-only change? Adding a comma after `}` would be the correct Flutter-style.

File: packages/flutter/test/widgets/list_view_test.dart [hunk 4]
```diff
@@ -120,6 +174,43 @@ void main() {
     log.clear();
   });
 
+  testWidgets('ListView large scroll jump and keepAlive (first child not keepAlive', (WidgetTester tester) async {
```
Comment (goderbauer) [inline packages/flutter/test/widgets/list_view_test.dart]: nit: missing closing `)`

File: packages/flutter/test/widgets/list_view_test.dart [hunk 5]
```diff
@@ -18,6 +18,60 @@ class TestSliverChildListDelegate extends SliverChildListDelegate {
...
+
+class _StatefulListViewState extends State<_StatefulListView> {
+  @override
+  Widget build(BuildContext context) {
+    return new GestureDetector(
+      onTap: () => setState,
+      child: new Directionality(
+        textDirection: TextDirection.ltr,
+        child: new ListView(
+          itemExtent: null,
```
Comment (goderbauer) [inline packages/flutter/test/widgets/list_view_test.dart]: nit: remove itemExtent as `null` is the default?

File: packages/flutter/test/widgets/list_view_test.dart [hunk 6]
```diff
@@ -18,6 +18,60 @@ class TestSliverChildListDelegate extends SliverChildListDelegate {
...
+  final WhetherToKeepAlive aliveCallback;
+  @override
+  _StatefulListViewState createState() => new _StatefulListViewState();
+}
+
+class _StatefulListViewState extends State<_StatefulListView> {
+  @override
+  Widget build(BuildContext context) {
+    return new GestureDetector(
+      onTap: () => setState,
```
Comment (goderbauer) [inline packages/flutter/test/widgets/list_view_test.dart]: Why this? No state seems to be changing?

File: packages/flutter/test/widgets/list_view_test.dart [hunk 7]
```diff
@@ -225,11 +316,11 @@ void main() {
...
-      new List<Widget>.generate(20, (int i) {
-        return new Container(
-          child: new Text('$i', textDirection: TextDirection.ltr),
-        );
-      })
+        new List<Widget>.generate(20, (int i) {
+          return new Container(
+            child: new Text('$i', textDirection: TextDirection.ltr),
+          );
+        })
```
Comment (goderbauer) [inline packages/flutter/test/widgets/list_view_test.dart]: nit: Flutter-style would be to add a comma after `}` and remove the newly introduced indentation.

File: packages/flutter/test/widgets/list_view_test.dart (hunk 6 above)
Comment (dnfield) [reply packages/flutter/test/widgets/list_view_test.dart]: Forces a rebuild. 

The issue is that on rebuild, an attempt at fixing this was dropping leading widgets with a keepAlive.  Rebuild here checks to make sure we're retaining those widgets.

Comment (goderbauer) [reply packages/flutter/test/widgets/list_view_test.dart]: Makes sense. Probably worth adding a comment here explaining the strange usage of setstate.

File: packages/flutter/lib/src/widgets/automatic_keep_alive.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/widgets/automatic_keep_alive.dart]: Done

File: packages/flutter/test/widgets/list_view_test.dart (hunk 6 above)
Comment (dnfield) [reply packages/flutter/test/widgets/list_view_test.dart]: Done

File: packages/flutter/lib/src/widgets/automatic_keep_alive.dart [hunk 8]
```diff
@@ -89,6 +89,10 @@ class _AutomaticKeepAliveState extends State<AutomaticKeepAlive> {
...
+          if (context == null) {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/automatic_keep_alive.dart]: use `!mounted`

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 9]
```diff
@@ -206,6 +206,28 @@ abstract class RenderSliverMultiBoxAdaptor extends RenderSliver
 
   @override
   void insert(RenderBox child, { RenderBox after }) {
+    // `after` might be coming in from a caller that didn't realize it was in the
+    // _keepAliveBucket, in which case it needs to be removed and reinserted
+    // to get connected to its siblings and be useful as an anchor.
+    if (after != null) {
+      final SliverMultiBoxAdaptorParentData afterParentData = after.parentData;
+      final int index = afterParentData.index;
+      if (_keepAliveBucket.containsKey(index)) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: rather than looking up the index in the map, i think you can just check _keptAlive on the child's parentData (then assert that it's in the map)

Comment (Hixie) [reply packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: (while you're at it, we should probably change _createOrObtainChild to do that too)

**Comment by Hixie** [comment]
This change seems wrong to me. Suppose you keep alive almost every child. When you try to insert a child that used to be after another child that is now itself kept alive, this code is going to insert every child back into the tree in an O(N^2) loop (each insert calls insert for every previous child, and does a loop over the children to find where to insert it).

Why are we trying to insert something after a child that is being kept alive? That seems like the real problem here. We're not supposed to ever be passing in an `after` to `insertAndLayoutChild` that isn't itself already laid out and ready.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart (hunk 9 above)
Comment (dnfield) [reply packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: _createOrObtainChild gets an index rather than a child.  Not sure that one makes sense to change to me.  But this one does make sense.

**Comment by dnfield** [comment]
> Suppose you keep alive almost every child. 

That's not a good idea for a sliverlist anyway - the whole point there is to be efficient with memory usage.  Keeping almost every child alive won't offer you that benefit.

> When you try to insert a child that used to be after another child that is now itself kept alive, this code is going to insert every child back into the tree in an O(N^2) loop (each insert calls insert for every previous child, and does a loop over the children to find where to insert it).

That is a worst case scenario, yes.

> Why are we trying to insert something after a child that is being kept alive? That seems like the real problem here. We're not supposed to ever be passing in an after to insertAndLayoutChild that isn't itself already laid out and ready.

The basic problem is that the `Sliver` widget tries to keep track of an anchor point (`_currentBeforeChild`).  It then tries to use that anchor point whenever someone wants to insert, so it doesn't have to go through the O(N^2) search.  That all works fine until that `_currentBeforeChild` got moved to the `KeepAlive` bucket by `SliverMultiBoxAdaptor` (and this happens for sure when there's a flick scroll or multi-tab jump).  `SliverMultiBoxAdaptor` has no way to tell `Sliver` that it messed up its cached child, and `Sliver` has no way to tell `SliverMultiBoxAdaptor` that it needs to find the siblings of its cached child.

The only way I can think of, short of the potentially costly search, is to have `SliverMultiBoxAdaptor` keep a `Map` of its `RenderBox`es, which would ruin the memory savings.

The bottom line seems to be that `KeepAlive` is costly in `Slivers`, and if you use it too much you're going to incur a lot of memory or processing overhead.

**Comment by dnfield** [comment]
One small enhancement might be to check if the `index` on the parent data is closer to `_firstChild.parentData.index` or `_lastChild.parentData.index` and potentially search backwards from the `_lastChild` - but it's still an expensive algorithm in the pathological case.

**Comment by dnfield** [comment]
macOS failure is a Dart VM garbage collection bug - https://github.com/dart-lang/sdk/issues/33111

**Comment by dnfield** [comment]
After offline conversation, this is the better fix @Hixie  was looking for.

- Ensure that the `_childElements` map is properly traversed as a sparse list and not inflated with garbage collected children
- Fixed test to actually ensure rebuild was actually happening with the tap. Test now fails with incorrect fix (where it was passing before).
- @Hixie  - a couple additions beyond what we discussed to correctly handle underflow and make sure we don't end up trying to concurrently iterate and modify `_childElements`.

**Comment by Steelsmasher** [comment]
Has this fix been implemented? Which version of flutter will I need?

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 10]
```diff
@@ -756,24 +756,24 @@ class SliverMultiBoxAdaptorElement extends RenderObjectElement implements Render
...
           _childElements[index] = newChild;
-          _currentBeforeChild = newChild.renderObject;
+          final SliverMultiBoxAdaptorParentData parentData = newChild.renderObject.parentData;
+          if (!parentData.keptAlive)
+            _currentBeforeChild = newChild.renderObject;
         } else {
           _childElements.remove(index);
         }
       }
+      // processElement may the Map - need to do a .toList() here.
```
Comment (SteveAlexander) [inline packages/flutter/lib/src/widgets/sliver.dart]: typo I think — "may the Map"

**Comment by dnfield** [comment]
@Steelsmasher it should be in master before much longer, and dev a little while after that.

File: packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart [hunk 11]
```diff
@@ -134,6 +134,10 @@ class SliverMultiBoxAdaptorParentData extends SliverLogicalParentData with Conta
   /// [RenderSliverMultiBoxAdaptor._keepAliveBucket].
   bool _keptAlive = false;
 
+  /// Whether the widget is currently being kept alive, i.e. has [keepAlive] set
+  /// to true and is offscreen.
+  bool get keptAlive => _keptAlive;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart]: normally we put the public getter above the private field and only have docs for the public getter. your call as to whether this case warrants an exception.

File: packages/flutter/lib/src/widgets/automatic_keep_alive.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/automatic_keep_alive.dart]: End comments with a period. Also since the comment is outside the if it should be a question ("Is the element still alive?"), or you can put it inside the block, or you can just remove it, the code speaks for itself here and the comment isn't really adding anything.

**Comment by Hixie** [comment]
LGTM
